### PR TITLE
Adding data transformers for preprocessing

### LINKED ISSUE #112: Implement some basic data transformations
It would be nice to begin supporting some basic data transformations. Things like scaling, etc. A model similar to sickit learn's transform interface would be a good jumping off point.

The following transformations would be nice for a start:

- Centering around zero with unit variance.
- Scaling Min/Max to some range.

To achieve this we should have a new trait defining our transform API. Something like this:

```rust
pub trait Transformer<T> {
    /// Transforms the inputs and stores the transformation in the Transformer
    fn transform(&mut self, inputs: T) -> Result<T, learning::Error>;

    /// Reverses the transform to recover the original data
    fn inv_transform(&self, inputs: T) -> Result<T, learning::Error>;
}
```

So our min max scaler may look like:

```rust
pub struct MinMaxScaler {
    /// Mins per column of input data
    input_min: Vec<usize>,
    /// Maxs per column of input data
    input_max: Vec<usize>,
    /// The min of the new data (default 0)
    scaled_min: f64,
    /// The max of the new data (default 1)
    scaled_max: f64,
}

impl Transformer<Matrix<f64>> for MinMaxScaler {
    fn transform(&mut self, inputs: Matrix<f64>) -> Result<Matrix<f64>, learning::Error> {
        // Find min and max of each column of features
        // For each column:
        //    Subtract the respective input_min and add the scaled_min
        //    Set each value x to x * (scaled_max / input_max)
    }

    fn inv_transform(&self, inputs: Matrix<f64>) -> Result<Matrix<f64>, learning::Error> {
        // Should fail if the Transformer has not been used to transform data yet
        //
        // Do the inverse of the transform:
        // For each column:
        //    Set each value x to x * (input_max / scaled_max)
        //    Subtract the scaled_min and add input_min
    }
}
```

The above is just to illustrate how this may look. Further requirements may mean something slightly different is needed. Some caveats we should be aware of:

- How to handle constant features? (If the std dev is zero we don't want to divide by this in scaling)
- How to handle class data? (i.e. values 0,1,2,3 representing one of 4 classes)

There are probably some other cases we should consider. But I think for now we can capture these as fail cases or just ignore them completely...

### PR DESCRIPTION
**Pull request by AtheMathmo**
This is a work in progress but I'm opening for feedback now. This will eventually resolve #112 .

In order to complete this PR I need to do the following:

- Implement a scaler for zero centering and unit variance.
- Add error handling in min/max case when column is a constant feature. I think for now just fail gracefully..?
- Add better documentation and review module layout.

### CODE REVIEW

File: src/data/transforms/minmax.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+/// which allows us to transform the input data to have a new minimum
+/// and maximum per column.
+///
+/// See the module description for more information.
+#[derive(Debug)]
+pub struct MinMaxScaler<T: Float> {
+    /// Mins per column of input data
+    scale_factors: Option<Vec<T>>,
+    /// Maxs per column of input data
+    const_factors: Option<Vec<T>>,
```
Comment (tafia) [inline src/data/transforms/minmax.rs]: Why not aligning comments and names (`min_factors` and `max_factors`) ?

Comment (AtheMathmo) [inline src/data/transforms/minmax.rs]: Thanks for reading this over.

This is an artifact of me changing the name of the field. I'll update now, thanks! (These comments wont be visible in docs as the fields aren't public - but still well worth getting right).

**Comment by AtheMathmo** [comment]
@tafia - does the rest of this PR look ok to you? If so I'll merge.

It doesn't feel quite right to me but is a good first draft for people to try out.

**Comment by tafia** [comment]
I am busy at the moment. I'll have another look tomorrow.
Feel free to merge I'll open an issue if needed.

**Comment by AtheMathmo** [comment]
No problem - I'll wait.

I'm away this weekend so wont be very responsive myself.

**Comment by AtheMathmo** [comment]
@tafia any other comments on this? If not I'll merge (if you don't have time yet but do want to take another look through that's fine - I can wait)

**Comment by tafia** [comment]
Sorry forgot it.
Looking now

File: src/data/transforms/minmax.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    scaled_min: T,
+    /// The max of the new data (default 1)
+    scaled_max: T,
+}
+
+/// Create a default `MinMaxScaler` with minimum of `0` and
+/// maximum of `1`.
+impl<T: Float> Default for MinMaxScaler<T> {
+    fn default() -> MinMaxScaler<T> {
+        MinMaxScaler {
```
Comment (tafia) [inline src/data/transforms/minmax.rs]: I would prefer using `MinMaxScaler::new` here. 
It is simpler if the `new` function ever changes ...

Comment (AtheMathmo) [inline src/data/transforms/minmax.rs]: I'm not sure I understand. There is a `new` function below this one - this is an implementation of the `Default` trait. I think it makes sense to have this here as I think scaling between 0 and 1 is common enough to warrant it as a default.

File: src/data/transforms/minmax.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                if !feature.is_finite() {
+                    return Err(Error::new(ErrorKind::InvalidData,
+                                          format!("Data point in column {} cannot be \
+                                                   processed",
+                                                  idx)));
+                }
+
+                if *feature < input_min[idx] {
+                    input_min[idx] = *feature;
+                }
```
Comment (tafia) [inline src/data/transforms/minmax.rs]: Very small point: 
You can avoid bound checks and `enumerate` using `zip` ... and only one zip if you work with tuples.
```rust
let mut input_min_max = vec![(T::min_value(), T::masx_value()); features];
```
... and computing the `scales` is simpler as well:
```rust
let scales = try!(input_min_max.iter().map(|(x, y)| {
    let s = (self.scaled_max - self.scaled_min) / (y - x);
    if s.is_finite() { 
        Ok(s) 
    } else {  
        Err(Error::new(ErrorKind::InvalidData,"Constant feature columns not supported."))
    }
}).collect::<Result<Vec<_>, _>>());
```

File: src/data/transforms/minmax.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        }
+
+        let consts = input_max.iter()
+            .zip(scales.iter())
+            .map(|(&x, &s)| self.scaled_max - x * s)
+            .collect::<Vec<_>>();
+
+        for row in inputs.iter_rows_mut() {
+            for i in 0..features {
+                row[i] = scales[i] * row[i] + consts[i];
```
Comment (tafia) [inline src/data/transforms/minmax.rs]: `utils::in_place_vec_bin_op` ?

File: src/data/transforms/minmax.rs (hunk 2 above)
Comment (tafia) [inline src/data/transforms/minmax.rs]:
```rust
impl<T: Float> Default for MinMaxScaler<T> {
    fn default() -> MinMaxScaler<T> {
        MinMaxScaler::new(T::zero(), T::one())
    }
}
```

Comment (AtheMathmo) [inline src/data/transforms/minmax.rs]: Ah I see - thanks!

**Comment by AtheMathmo** [comment]
I pushed some updates from your feedback. Would you mind taking another look?

**Pull request merged**